unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -106,92 +105,126 @@ public int length() {
    */
   public UTF8String substring(final int start, final int until) {
-    if (until <= start || start >= bytes.length) {
+    if (until <= start || start >= size) {
       return UTF8String.fromBytes(new byte[0]);
     }
 
     int i = 0;
     int c = 0;
-    for (; i < bytes.length && c < start; i += numBytes(bytes[i])) {
+    for (; i < size && c < start; i += numBytes(getByte(i))) {
       c += 1;
     }
 
     int j = i;
-    for (; j < bytes.length && c < until; j += numBytes(bytes[i])) {
+    for (; j < size && c < until; j += numBytes(getByte(i))) {
       c += 1;
     }
 
-    return UTF8String.fromBytes(Arrays.copyOfRange(bytes, i, j));
+    byte[] bytes = new byte[j - i];
+    copyMemory(base, offset + i, bytes, BYTE_ARRAY_OFFSET, j - i);
+    return UTF8String.fromBytes(bytes);
   }
 
   public boolean contains(final UTF8String substring) {
-    final byte[] b = substring.getBytes();
-    if (b.length == 0) {
+    if (substring.size == 0) {
       return true;
     }
 
-    for (int i = 0; i <= bytes.length - b.length; i++) {
-      if (bytes[i] == b[0] && startsWith(b, i)) {
+    byte first = substring.getByte(0);
+    for (int i = 0; i <= size - substring.size; i++) {
+      if (getByte(i) == first && matchAt(substring, i)) {
         return true;
       }
     }
     return false;
   }
 
-  private boolean startsWith(final byte[] prefix, int offsetInBytes) {
-    if (prefix.length + offsetInBytes > bytes.length || offsetInBytes < 0) {
+  private long getLong(int i) {
+    return UNSAFE.getLong(base, offset + i);
+  }
+
+  private byte getByte(int i) {
+    return UNSAFE.getByte(base, offset + i);
+  }
+
+  private boolean matchAt(final UTF8String s, int pos) {
+    if (s.size + pos > size || pos < 0) {
       return false;
     }
+
     int i = 0;
-    while (i < prefix.length && prefix[i] == bytes[i + offsetInBytes]) {
-      i++;
+    while (i <= s.size - 8) {
+      if (getLong(pos + i) != s.getLong(i)) {
+        return false;
+      }
+      i += 8;
     }
-    return i == prefix.length;
+    while (i < s.size) {
+      if (getByte(pos + i) != s.getByte(i)) {
+        return false;
+      }
+      i += 1;
+    }
+    return true;
   }
 
   public boolean startsWith(final UTF8String prefix) {
-    return startsWith(prefix.getBytes(), 0);
+    return matchAt(prefix, 0);
   }
 
   public boolean endsWith(final UTF8String suffix) {
-    return startsWith(suffix.getBytes(), bytes.length - suffix.getBytes().length);
+    return matchAt(suffix, size - suffix.size);
   }
 
   public UTF8String toUpperCase() {
+    // this is locale aware
     return UTF8String.fromString(toString().toUpperCase());
   }
 
   public UTF8String toLowerCase() {
+    // this is locale aware
     return UTF8String.fromString(toString().toLowerCase());
   }
 
   @Override
   public String toString() {
     try {
-      return new String(bytes, "utf-8");
+      // this is slow
+      return new String(getBytes(), "utf-8");
```

Comment:
`String` actually has the constructor like:
```
public String(byte bytes[], int offset, int length, String charsetName)
```
Will that be helpful to avoid the possible byte array copying inside of the `getBytes`?